src/mon/command_multiplexer.h

```diff
@@ -0,0 +1,312 @@
+template <typename CommandType>
+struct command_registry
+{
+ using command_type = CommandType;
+
+ // The type returned by an operation:
+ using status_type  = std::optional<typename command_type::result_type>;
+
+ private:
+ command_registry(const command_registry&)              = delete;
+ command_registry& operator=(const command_registry&)   = delete;
+
+ public:
+ const std::string name;      // eg. the name of the registry 
+
+ std::vector<std::unique_ptr<command_type>> commands;
+
+ public:
+ command_registry(std::string_view name_)
+  : name(name_)
+ {}
+
+ virtual ~command_registry() = default;
+
+ public:
+ auto find_cmd_for_tag(std::string_view candidate_tag) const
+ {
+    using std::end;
+    using std::begin;
+
+    auto tag_match = [candidate_tag](const auto& command) {
+        auto tags = command->tags();
+        return end(tags) != std::find(begin(tags), end(tags), candidate_tag);
+    };
+
+    return std::find_if(begin(commands), end(commands), tag_match); 
+ }
+
+ // True if a "tag" is found:
+ bool in_registry(std::string_view tag) const
+ {
+    return end(commands) != find_cmd_for_tag(tag);
+ }
+
+ // True if any tag in the "tags" sequence is found:
+ template <typename K, typename A, template <typename, typename> typename SeqT = std::vector>
+ bool in_registry(const SeqT<K, A>& tags) const
+ {
+    return std::end(tags) != std::find_if(std::begin(tags), std::end(tags), 
+                                         [this](const auto& tag) {
+                                           return this->in_registry(tag);
+                                        });
+ }
+
+ void add_command(std::unique_ptr<CommandType> cmd) 
+ {
+    using namespace std;
+
+    auto tags(cmd->tags());
+
+    if (in_registry(tags)) {
+      ostringstream os;
+      os << "already in registry: ";
+      copy(begin(tags), end(tags), experimental::ostream_joiner(os, ','));
+      throw runtime_error(os.str());
+    }
+
+    commands.push_back(std::move(cmd));
+ }
+
+ // Actually accept a user's function and call it, or decline service:
+ public:
+ status_type apply(const operation_type op_type, 
+                   std::string_view cmd_prefix, 
+                   const typename CommandType::state_type& state)
+ {
+    auto cmd = find_cmd_for_tag(cmd_prefix);
+
+    // Command not found:
+    if (std::end(commands) != cmd) 
+        return {};
+
+    auto& c = *cmd;
+
+    auto [result, tag] = c->accepts(op_type, cmd_prefix, c->tags());
+
+    // Command found, but did not accept:
+    if (!result)
+        return {};
+
+    if (ceph::operation_type::preprocess == op_type)
+     return { c->at_preprocess(state, cmd_prefix.substr(tag.size())) };
+
+    if (ceph::operation_type::prepare == op_type)
+     return { c->at_prepare(state, cmd_prefix.substr(tag.size())) };
+
+    // If this happens, something's very wrong:
+    std::ostringstream os;
+    os << "invalid operation type " << static_cast<int>(op_type);
+    throw std::invalid_argument(os.str());
```

Comment:
It will definitely be great to have Contracts for exactly this sort of thing! Indeed, this error "shouldn't ever happen" and I also doubt the system could really handle it. One consideration here before committing (hah-hah) to assert(): it would be nice to see the runtime value that led to this condition. So one idea would be to throw here and catch invalid_argument() in the actual handling code, which could log the error and... assert(). I don't know if that's actually preferable (and also Ceph does some fiddling with the error and termination handlers, so maybe we have some customized behavior). Is it better to just lose that and drop out here? I'm ok either way.